File: medium/0098_validate_binary_search_tree/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <stack>

class Solution {
public:
  bool isValidBST(utils::TreeNode *root) {
    if (!root)
      return true;

    std::stack<std::tuple<utils::TreeNode *, long, long>> s;
    s.push({root, INT64_MIN, INT64_MAX});

    while (!s.empty()) {
      auto [node, left, right] = s.top();
      s.pop();

      if (left >= node->val || node->val >= right)
        return false;

      if (node->left)
        s.push({node->left, left, node->val});
      if (node->right)
        s.push({node->right, node->val, right});
    }

    return true;
  }
};

File: medium/0098_validate_binary_search_tree/solution.py
from typing import Optional

from utils.python.nodes import TreeNode


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        stack = [(root, float("-inf"), float("inf"))]
        while stack:
            node, left, right = stack.pop()

            if not (left < node.val < right):
                return False

            if node.left:
                stack.append((node.left, left, node.val))
            if node.right:
                stack.append((node.right, node.val, right))

        return True
